### DISCUSSION TITLE
Database connection and automation

### DISCUSSION THREAD

**Discussion by almitte** [DependencyTrack/dependency-track#1475] [Q&A]
Hey,

I am pretty new to Dependency-Track and IT-projects in general. I want to connect a Database to D-Track to save the SBoMs and automatically upload them to D-Track. I have read the instruction for that on the website, but am still unsure.

-  Where can I change the database settings as shown in the documentation exactly? 

- Can I automate the upload of SBoMs through the use of that database and if so what is the optimal way to setup such a database?

- If not is there any option to automate it?

Any help or documentation would be great. Thanks in advance.

**Comment by stevespringett** [comment]
Refer to:
https://docs.dependencytrack.org/getting-started/deploy-docker/
https://docs.dependencytrack.org/usage/cicd/
https://docs.dependencytrack.org/integrations/community-integrations/

**Reply by almitte** [reply]
So i just replace the Database Settings in the .yml File with the ones given at https://docs.dependencytrack.org/getting-started/database-support/ ? Just for clarification, the database stores the uploaded SBoMs to Dependency-Track? 

I also found the link to the github action and Jenkins plugin, thanks!

**Reply by stevespringett** [reply]
> Just for clarification, the database stores the uploaded SBoMs to Dependency-Track?

Currently, DT does not store BOMs. It consumes and imports them, analyzes the data, and can dynamically export a BOM on demand from the current inventory, but it does not currently save BOMs. Ticket #877 will introduce a BOM retention policy and the ability to store/retrieve BOMs by their serial number.

**Reply by almitte** [reply]
Our organisation is at the consumer part of the supply chain. We just get the software and related BOMs, want to save and analyze them. Most solutions like Jenkins seems to be more fitted for software delevopers...

1) What does the SQL database exactly do/save when I connect it to DT when it _doesnt save the BOMs?_

2) Is there any other option to automatically upload the BOMs to DT from a data storage (and keep them up to date if a new one will be available)? Or would I need to code a script that gets the BOMs and uploads them via the API? 

Thank you so much in advance for the help.

**Reply by stevespringett** [reply]
> What does the SQL database exactly do/save when I connect it to DT when it doesnt save the BOMs?

It decomposes the BOM into discrete inventory so that each component and/or service can be analyzed independently. DT stores all the inventory information in the database, all project information for how the bom is associated with a project, all public vulnerability information, advisories, lastest versions of components, policy information, etc. 

> Is there any other option to automatically upload the BOMs to DT from a data storage (and keep them up to date if a new one will be available)? Or would I need to code a script that gets the BOMs and uploads them via the API?

I'm not sure I understand the question. You can automate uploads via API - and there are many tools that do this as well. And there are manual use cases, where you can upload via the UI. So automated and manual use cases are supported. Are you looking for something else?

**Comment by almitte** [comment]
Thank you for your patience and help. 
Maybe I describe the use case further:
Our organisations get the created SBoMs from our suppliers. My idea was that we first save the Json files in a central repository/data storage. Afterwards I want to automatically load the SBoMs from that Data storage into Dtrack. There the decomposed SBoMs will than be stored on a SQL Server which I already set up.

1) My question was wether there are any options (and If so which?) to automatically get the json Files from our data storage and upload them via the API to dtrack or if we have to code this on our own.

2) Is this even a viable workflow or would you recommend another workflow from the provision of the Sboms by our suppliers to the upload into dtrack. It seems to be somehow redundant to save the Json files in a data storage and than the decomposed SBoMs in the SQL Server.

Thanks so much in advance.

**Reply by stevespringett** [reply]
You could use the [CycloneDX BOM Repo Server](https://github.com/CycloneDX/cyclonedx-bom-repo-server) to publish & store BOMs. A future version of DT will likely include it. So there is a means to store BOMs. The only thing missing for your use case is to glue the repo server and DT together with some simple API calls.

> It seems to be somehow redundant to save the Json files in a data storage and than the decomposed SBoMs in the SQL Server.

No, not at all. There are many use cases for having them independent from one another. A future version of DT will be including a configurable retention policy that will likely use the BOM Repo Server as a storage service.
